### PR TITLE
👌 IMPROVE: `Repository.hash()` speed

### LINKED ISSUE #4918: `Repository.hash` seems very inefficient
https://github.com/aiidateam/aiida-core/blob/5f87661e9099b1e8875254df8b133dd43b58c5bc/aiida/repository/repository.py#L96-L110

@giovannipizzi @sphuber Why do we need to read every file and regenerate their hash?
why not just re-use the hashes of the files that are already stored e.g.:

```python
    def hash(self) -> str:
        """Generate a hash of the repository's contents.

        :return: the hash representing the contents of the repository.
        """
        objects = {}
        for root, dirnames, filenames in self.walk():
            objects['__dirnames__'] = dirnames
            for filename in filenames:
                objects[str(root / filename)] = self.get_object(root / filename).key

        return make_hash(objects)

```

### PR DESCRIPTION
**Pull request by chrisjsewell**
closes #4918

Uses `hashlib.sha256` by default then, to be super safe, only uses the key for the object-store if `self.container.hash_type == 'sha256'`

before:

```
--------------------------------------------- benchmark 'node': 1 tests ---------------------------------------------
Name (time in ms)              Min       Max     Mean  StdDev   Median     IQR  Outliers      OPS  Rounds  Iterations
---------------------------------------------------------------------------------------------------------------------
test_store_with_object     10.4047  105.4098  14.0552  9.3613  13.0171  1.9786       1;4  71.1483     100           1
---------------------------------------------------------------------------------------------------------------------
```

after:

```
------------------------------------------- benchmark 'node': 1 tests -------------------------------------------
Name (time in ms)             Min     Max    Mean  StdDev  Median     IQR  Outliers       OPS  Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------
test_store_with_object     5.8858  9.1815  6.8171  0.7261  6.5448  0.9458      22;3  146.6907     100           1
-----------------------------------------------------------------------------------------------------------------
```

### CODE REVIEW

File: aiida/repository/backend/abstract.py
Comment (giovannipizzi) [inline aiida/repository/backend/abstract.py:122]: This might exhaust RAM if the object is very big (possible).

You should compute the hash by reading the file in chunks.
You can copy-paste and adapt from here as an example:
https://github.com/aiidateam/disk-objectstore/blob/a2561a40cf2e9a4d58325387f913ccf08111f5fd/disk_objectstore/utils.py#L815-L839

(It's possible we've the same issue elsewhere in AiiDA, but I would start improving those one by one)

**Review by giovannipizzi — CHANGES_REQUESTED** [review]
Seems a very good and safe approach.

Just a comment to avoid memory issues with big objects (feel free to dismiss my review once fixed, is @sphuber approves)

File: aiida/repository/backend/abstract.py
Comment (chrisjsewell) [reply aiida/repository/backend/abstract.py:122]: cheers will do

File: aiida/repository/backend/disk_object_store.py
Comment (sphuber) [inline aiida/repository/backend/disk_object_store.py:116]: I was struggling for a bit to understand why this conditional was necessary because the container always uses the hash of the content as the key, so it shouldn't matter what hash type is used. However, then I realized that this would be a problem higher up that expects the same hash of a file regardless of what type of repository it is stored in. I think this is not directly obvious and think this should probably be written in the `Repository.get_hash` method docstring.

This may cause problems in the future when another repo is introduced that uses a different hashing type and all of a sudden caching and other things that rely on hashes may start failing, no?

Comment (giovannipizzi) [reply aiida/repository/backend/disk_object_store.py:116]: Yes, indeed might not be obvious that this function *must* return a `sha256` hash (as this is what AiiDA expects, and should not change between different backends). Would be good to clarify.

File: aiida/repository/backend/abstract.py
Comment (chrisjsewell) [reply aiida/repository/backend/abstract.py]: Implemented `aiida.common.hashing.chunked_file_hash`

File: aiida/repository/backend/disk_object_store.py
Comment (chrisjsewell) [reply aiida/repository/backend/disk_object_store.py:117]: Changed docstring to assert that SHA-256 hashing must be returned.

> This may cause problems in the future when another repo is introduced that uses a different hashing type

This would only now happen if the repo class specifically overrode `AbstractRepositoryBackend.get_object_hash` and did not "respect" the method's docstring, which would then be a bug in that classes implementation

**Review by giovannipizzi — APPROVED** [review]
Thanks!

**Review by sphuber — APPROVED** [review]
All good thanks @chrisjsewell

**Pull request merged**
